internal/iam/query.go

```diff
@@ -453,6 +453,195 @@ const (
+       where principal_id in (select id
+                                from user_groups)
+    ),
+    user_roles (role_id) as (
+      select role_id
+        from iam_user_role
+       where principal_id in (select id
+                                from users)
+    ),
+    all_associated_roles (role_id) as (
+      select role_id
+        from group_roles
+       union
+      select role_id
+        from user_roles
+       union
+      select role_id
+        from managed_group_roles
+    ),
+    roles_with_grants (role_id, canonical_grant) as (
+      select iam_role_grant.role_id,
+             iam_role_grant.canonical_grant
+        from iam_role_grant
+        join iam_role
+          on iam_role.public_id = iam_role_grant.role_id
+        join iam_grant
+          on iam_grant.canonical_grant = iam_role_grant.canonical_grant
+       where iam_role.public_id in (select role_id
+                                      from all_associated_roles)
+         and iam_grant.resource = any(@resources)
+    ),
+    global_roles_with_descendant_grant_scopes as (
+      select iam_role_global.public_id         as role_id,
+             iam_role_global.scope_id          as role_scope_id,
+             'global'                          as role_parent_scope_id,
+             iam_role_global.grant_scope       as grant_scope,
+             roles_with_grants.canonical_grant as canonical_grant
+        from iam_role_global
+        join roles_with_grants
+          on roles_with_grants.role_id = iam_role_global.public_id
+       where iam_role_global.grant_scope = 'descendants'
+    ),
+    global_roles_with_individual_grant_scopes as (
+      select iam_role_global.public_id         as role_id,
+             iam_role_global.scope_id          as role_scope_id,
+             'global'                          as role_parent_scope_id,
+             individual.scope_id               as grant_scope,
+             roles_with_grants.canonical_grant as canonical_grant
+        from iam_role_global
+        join roles_with_grants
+          on roles_with_grants.role_id = iam_role_global.public_id
+        join iam_role_global_individual_project_grant_scope individual
+          on individual.role_id = iam_role_global.public_id
+       where individual.scope_id = @request_scope
+    ),
+    org_roles_with_children_grant_scopes as (
+      select iam_role_org.public_id            as role_id,
+             iam_role_org.scope_id             as role_scope_id,
+             'global'                          as role_parent_scope_id,
+             iam_role_org.grant_scope          as grant_scope,
+             roles_with_grants.canonical_grant as canonical_grant
+        from iam_role_org
+        join roles_with_grants
+          on roles_with_grants.role_id = iam_role_org.public_id
+        join iam_scope_org
+          on iam_scope_org.scope_id = iam_role_org.scope_id
+        join iam_scope_project
+          on iam_scope_project.parent_id = iam_scope_org.scope_id
```

Comment:
You might not need to join with `iam_scope_org` here

```suggestion
        join iam_scope_project
          on iam_scope_project.parent_id = iam_role_org.scope_id
```